relay/network_graph/alg.py

```diff
@@ -0,0 +1,129 @@
+"""graph algorithms"""
+
+import heapq
+from typing import List, Set, Dict, Callable
+import networkx as nx
+import abc
+
+
+class CostAccumulator(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def zero(self):
+        """return 'zero cost' element, which is the initial cost for a one-node path
+
+        This must be less than or equal to any value
+        accumulate_cost_for_next_hop returns.
+        """
+        pass
+
+    @abc.abstractmethod
+    def accumulate_cost_for_next_hop(
```

Comment:
Took me some time to understand this, what costs come back, what is graph data, maybe we can rename it and name the arguments better?
for example what about 
`total_cost_from_start_to_dst(self, cost_from_start_to_node, node, dst, edge_data_node_dst)`